### PR TITLE
feat: Improvements to GS1 imports to prepare integration of Agena3000

### PR DESCRIPTION
**Pull request by stephanegigandet**
This PR includes:

- changes to use the Perl XML::XML2JSON module instead of nodejs xml2json module, in order to remove the dependency on nodejs (as suggested by @alexgarel )
- support for GS1 messages that contain nested products (a pallet that contains cartons that contain the actual products sold to consumers)
- updated expected test results: as we might have more than 1 product in a GS1 message, each test now returns an array of product data
- new test cases with sample data from Agena3000 in XML converted by the Perl XML::XML2JSON module

### CODE REVIEW

File: scripts/convert_gs1_xml_to_json_in_dir.js [hunk 1]
```diff
@@ -1,12 +1,14 @@
 // This script is used to convert GDSN data from Equadis in XML format
 // to a corresponding JSON structure
 
-const xml2json = require('xml2json')
-const fs = require("fs")
+const process = require('process')
```
Comment (alexgarel) [inline scripts/convert_gs1_xml_to_json_in_dir.js:4]: needs a ";"

linting does not pass, see https://github.com/openfoodfacts/openfoodfacts-server/runs/5869772770?check_suite_focus=true#step:5:39

File: lib/ProductOpener/GS1.pm [hunk 2]
```diff
@@ -1208,7 +1235,19 @@ sub gs1_to_off ($$$) {
 				elsif (ref($json_ref->{$source_field}) eq "ARRAY") {
 					foreach my $json_array_entry_ref (@{$json_ref->{$source_field}}) {
 
-						gs1_to_off($source_target, $json_array_entry_ref, $results_ref);
+						# We should have an array of hashes, but in some CodeOnline files we have an array with an empty hash..
```
Comment (alexgarel) [inline lib/ProductOpener/GS1.pm:1238]:
```suggestion
						# We should have an array of hashes, but in some CodeOnline files we have an array with an empty array..
```

File: lib/ProductOpener/GS1.pm [hunk 3]
```diff
@@ -1217,31 +1256,110 @@ sub gs1_to_off ($$$) {
...
+One GS1 message can include 1 or more products, typically products that contain other products
+(e.g. a pallet of cartons of products).
+
+=head3 Arguments
 
-=head3 Return value
+=head4 $json_ref Reference to a decoded JSON structure
 
-=head4 Reference to a hash of fields
+=head4 Reference to an array of product data
```
Comment (alexgarel) [inline lib/ProductOpener/GS1.pm:1340]:
```suggestion
=head4 $product_ref - Reference to an array of product data
```

File: scripts/convert_gs1_xml_to_json_in_dir.pl [hunk 4]
```diff
@@ -0,0 +1,71 @@
+#!/usr/bin/perl -w
+
```
Comment (alexgarel) [inline scripts/convert_gs1_xml_to_json_in_dir.pl:2]:
```suggestion

# Perl script to convert XML received from GS1 to JSON files, more easily processed

```

File: scripts/equadis-import/run_equadis_import.sh [hunk 5]
```diff
@@ -12,11 +12,10 @@ find /home/sftp/equadis/data/ -mtime -2 -type f -exec cp {} /srv2/off-pro/equadi
 
 export NPM_CONFIG_PREFIX=~/.npm-global
 
-node equadis-xml2json.js
+node /srv/off-pro/scripts/convert_gs1_xml_to_json_in_dir.js /srv2/off-pro/equadis-data-tmp/
```
Comment (alexgarel) [inline scripts/equadis-import/run_equadis_import.sh:15]: why not use the new .pl here ?

**Review by alexgarel — APPROVED** [review]
LGTM

kudos

File: scripts/equadis-import/run_equadis_import.sh (hunk 5 above)
Comment (stephanegigandet) [reply scripts/equadis-import/run_equadis_import.sh:15]: I want to test it out more thoroughly first, to check that we end up with the exact same results once the products are imported into OFF.

File: scripts/convert_gs1_xml_to_json_in_dir.js (hunk 1 above)
Comment (stephanegigandet) [reply scripts/convert_gs1_xml_to_json_in_dir.js]: fixed

**Pull request merged**